staging/src/k8s.io/apiserver/pkg/server/filters/waitgroup.go

```diff
@@ -41,21 +47,32 @@ func WithWaitGroup(handler http.Handler, longRunning apirequest.LongRunningReque
 			return
 		}
 
-		if !longRunning(req, requestInfo) {
-			if err := wg.Add(1); err != nil {
-				// When apiserver is shutting down, signal clients to retry
-				// There is a good chance the client hit a different server, so a tight retry is good for client responsiveness.
-				w.Header().Add("Retry-After", "1")
-				w.Header().Set("Content-Type", runtime.ContentTypeJSON)
-				w.Header().Set("X-Content-Type-Options", "nosniff")
-				statusErr := apierrors.NewServiceUnavailable("apiserver is shutting down").Status()
-				w.WriteHeader(int(statusErr.Code))
-				fmt.Fprintln(w, runtime.EncodeOrDie(scheme.Codecs.LegacyCodec(v1.SchemeGroupVersion), &statusErr))
+		if longRunning(req, requestInfo) {
+			handler.ServeHTTP(w, req)
+			return
+		}
+
+		if err := wg.Add(1); err != nil {
+			// shutdown delay duration has elapsed and SafeWaitGroup.Wait has been invoked
+			// we are going to allow execution of the requests that WithRetryAfter is
+			// excluding from sending a Retry-After response.
+			if excludeFn(req) {
```

Comment:
why is this not like long-running? Can we combine both predicates?

i.e. a function called excludeFunc in withWaitGroup and then it is not really excluded.

Reply:
we don't want to combine with `longRunning` - `excludeFn` is relevant when shutdown delay duration has elapsed. Only then `WithRetryAfter` starts responding with retry-after response. 
inside `withWaitGroup` we are not sending retry-after for the white-listed (allowed) requests and we are not counting them either.